### PR TITLE
DAOS-691 tests: read an array while rebuild is in progress

### PR DESCRIPTION
**Pull request by phender**
Adding a new functional test to write a dkey, fail a target, read
the dkey with a recovery in progress.

Signed-off-by: Phil Henderson <phillip.henderson@intel.com>

### CODE REVIEW

File: src/tests/ftest/util/io_utilities.py [hunk 1]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
+
+    Returns:
+        None
+
+    Raises:
+        DaosTestError: if the rebuild completes before the read is complete
+            or read errors are detected
+
+    """
+    x = 0
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/io_utilities.py [hunk 2]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
+
+    Raises:
+        DaosTestError: if an error is detected writing the objects or
+            verifying the write of the objects
+
+    """
+    if log:
+        log.info("Creating objects in the container")
+    object_list = []
+    for x in range(obj_qty):
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/io_utilities.py [hunk 3]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/io_utilities.py [hunk 4]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
+                log.info(
+                    "<obj: %s, rec: %s>: Passed reading data "
+                    "(dkey=%s, akey=%s)",
+                    x, y, record_info["dkey"], record_info["akey"])
+
+            # Read the next record in this object or the next object
+            y += 1
+            if y >= len(written_objects[x]["record"]):
+                x += 1
+                y = 0
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/io_utilities.py [hunk 5]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
+            else:
+                log.info(
+                    "<obj: %s, rec: %s>: Passed reading data "
+                    "(dkey=%s, akey=%s)",
+                    x, y, record_info["dkey"], record_info["akey"])
+
+            # Read the next record in this object or the next object
+            y += 1
+            if y >= len(written_objects[x]["record"]):
+                x += 1
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/io_utilities.py [hunk 6]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
+                    x, y, record_info["dkey"], record_info["akey"],
+                    read_data, record_info["data"])
+            else:
+                log.info(
+                    "<obj: %s, rec: %s>: Passed reading data "
+                    "(dkey=%s, akey=%s)",
+                    x, y, record_info["dkey"], record_info["akey"])
+
+            # Read the next record in this object or the next object
+            y += 1
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/io_utilities.py [hunk 7]
```diff
@@ -125,3 +142,256 @@ def write_quantity(container, size_in_bytes):
...
+    Returns:
+        None
+
+    Raises:
+        DaosTestError: if the rebuild completes before the read is complete
+            or read errors are detected
+
+    """
+    x = 0
+    y = 0
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/io_utilities.py]: fixing

File: src/tests/ftest/util/general_utils.py [hunk 8]
```diff
@@ -90,7 +101,205 @@ def process_host_list(hoststr):
...
+@fail_on(DaosApiError)
+def get_pool_status(pool, log):
+    """Determine if the pool rebuild is complete.
+
+    Args:
+        pool (DaosPool): pool for which to determine if rebuild is complete
+        log (logging): logging object used to report the pool status
+
+    Returns:
+        None
```
Comment (phender) [inline src/tests/ftest/util/general_utils.py]: This function returns the return value of the pool.pool_query() function - need to update comment.

File: src/tests/ftest/rebuild/read_array.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        except DaosTestError as error:
+            # if pool_svcn
+            self.log.warning(
+                "Detected exception while reading during rebuild: %s",
+                str(error))
+
+        # Confirm rebuild completes
+        wait_for_rebuild(self.pool, self.log, False, 1)
+        self.log.info("Rebuild completion detected")
+
```
Comment (wangdi1) [inline src/tests/ftest/rebuild/read_array.py]: Maybe you should verify every replica of the array object after rebuild. Ah, we do not have such api for python yet?

Comment (phender) [reply src/tests/ftest/rebuild/read_array.py]: I'll add a story to update this test when the API is available

**Pull request merged**